add support for distro release filtering

iurly opened the pull request:
add an extra parameter to add the possibility to filter
packages for a given distro release.
The main use case is ubuntu, whose list of packages
(considering all releases) is simply humongous.

This way we can filter by adding -R "impish".

gnosek commented on probe_builder/kernel_crawler/__init__.py:
```diff
@@ -32,7 +32,9 @@
+    dist = DISTROS[distro]()
+    dist.set_distro_filter(distro_filter)
+    pt = dist.get_package_tree(kernel_filter)
+    dist.set_distro_filter('')
```
Do we need this line? AIUI, we're creating a new instance and `set_distro_filter` sets only an instance field so it should be safe to just let `dist` go out of scope without resetting it.

If I'm wrong, I'm happy to be enlightened :)

iurly replied:
Oh yes, you're right, that's not strictly needed. I just put it there for clarity.

gnosek replied:
For me personally, it decreases the clarity a little since it seems to suggest we're holding onto the `dist` object longer than we are

deepskyblue86 replied:
I'll argue on the clarity here :sweat_smile: 
This extra like makes me think it's needed while it's not, so the effect is more :thinking:  to me

gnosek reviewed (commented):
I'm not a fan of setting `distro_filter` as an instance field, for two reasons:
* it's harder to reason about the lifetime of the setting when we set/reset it
* it introduces an asymmetry with the kernel filter, which is passed around as a parameter

Other than that, I love it!

iurly commented:
> I'm not a fan of setting `distro_filter` as an instance field, for two reasons:
> 
>     * it's harder to reason about the lifetime of the setting when we set/reset it
> 
>     * it introduces an asymmetry with the kernel filter, which is passed around as a parameter
> 
> 
> Other than that, I love it!

The first implementation was indeed a parameter just like the kernel filter.
The reason I didn't like it was that it changed the method signature for lots of sub classes, and it was then easy to forget to update some... especially the ones that came in through some pending PRs.
The instance field was the first thing that came to mind which didn't require to change every single distro.
Shall we make a generic class like a `CrawlerFilter` (carrying information about both distro & kernel filter, plus whatever we'll want to come up with in the future) instead?

gnosek commented:
> Shall we make a generic class like a `CrawlerFilter` (carrying information about both distro & kernel filter, plus whatever we'll want to come up with in the future) instead?

Yes please!

deepskyblue86 commented on probe_builder/kernel_crawler/repo.py:
```diff
@@ -19,27 +19,39 @@ def to_s(s):
             for mirror in mirrors:
+                mirror.set_distro_filter(self.distro_filter)
                 repos.extend(mirror.list_repos())
+                mirror.set_distro_filter('')
```
do we need to reset it here?

iurly committed:
distro filtering: rework so to have a global filtering object

gnosek commented on probe_builder/kernel_crawler/repo.py:
```diff
@@ -3,9 +3,15 @@
 import click
 import sys
 
+class CrawlerFilter(object):
+    def __init__(self, distro_filter='', kernel_filter=''):
+        self.distro_filter = distro_filter
+        self.kernel_filter = kernel_filter
+
+EMPTY_FILTER=CrawlerFilter()
```
Since we're using `EMPTY_FILTER` as a default arg, we don't want it to be accidentally modified. Can we make this class a named tuple? I think this wouldn't require any changes at the call site, except that you can't modify the fields after you create an instance (which I don't think we do?)

iurly committed:
make CrawlerFilter a namedtuple

iurly committed:
fix rpm filtering

iurly committed:
add support for distro release filtering

add an extra parameter to add the possibility to filter
packages for a given distro release.
The main use case is ubuntu, whose list of packages
(considering all releases) is simply humongous.

This way we can filter by adding -R "impish".
